Add Transmission Connector

itshoro opened the pull request:
This PR adds support for a *transmission*-connector.

It implements the **add**-, **get**- and  **remove**-actions, which are present in the legacy Yousei implementation.

There are some things I'm not completely sure about.

https://github.com/WiiPlayer2/Yousei/blob/02a117f47ebf14678b5e60841970290d3b1cf858/Yousei.Connectors/Transmission/AddAction.cs#L26

Is calling `ConfigureAwait(false)` necessary here?

https://github.com/WiiPlayer2/Yousei/blob/02a117f47ebf14678b5e60841970290d3b1cf858/Yousei.Connectors/Transmission/RemoveAction.cs#L22

This function isn't async, is there a better way to do this?

**At this point we would also still need to add the connector**, by callling `YouseiReloaded.Internal.ConnectorRegistry.Register(new TransmissionConnector())`, however I also am unsure on how and especially **where** I should call this.

itshoro committed:
Add TransmissionConnector

- Implement AddAction / AddArguments
- Implement GetAction / GetArguments
- Implement RemoveAction / RemoveArguments
- Implement TransmissionConnection
- Implement TransmissionConnector, it needs to pass  a transmission client to all supported actions

itshoro committed:
Add needed Dependencies

- Add Reference for the Yousei.Core and Yousei.Shared projects
- Add System.Reactive
- Add Transmission.API.RPC

WiiPlayer2 commented:
In library code `ConfigureAwait(false)` should almost always be used. I chose to ignore this for now as we are not using any synchronization context so the call doesn't matter.

If the transmission api method is not returning a task then there is nothing we can do about it. We'll just leave it be for now until there is an update which makes it return a task.

All static connectors should be registered in the concrete implementation namely inside the constructor of `YouseiReloaded.Internal.ConnectorRegistry`.

WiiPlayer2 commented on Yousei.Connectors/Transmission/GetAction.cs:
```diff
@@ -0,0 +1,25 @@
+            var fields = await arguments.Fields.Resolve<string[]>(context);
+            var ids = await arguments.Ids.Resolve<int[]>(context);
+
+            await client.TorrentGetAsync(fields, ids);
```
The information returned by this should be saved in the context.

itshoro replied:
Should this also apply to `client.TorrentAddAsync()`?

WiiPlayer2 replied:
If it returns a meaningful object (which it does if I recall correctly) then yes, this should also apply to `TorrentAddSync`.

WiiPlayer2 commented on Yousei.Connectors/Transmission/TransmissionConnector.cs:
```diff
@@ -0,0 +1,19 @@
+        private readonly TransmissionConnection connection;
+
+        public TransmissionConnector(Client client) : base("transmission")
+        {
```
The `TransmissionConnector` should not get a `Transmission.API.RPC.Client` and rather create it as part of its `GetConnection` method. The current implementation would disallow connections to different transmission servers and force the caller to connect to the server which goes against the architecture.
There should be configuration class containing all relevant connection arguments (api key etc.) and then those should be used in `GetConnection` to create a `TransmissionConnection` with a `Transmission.API.RPC.Client` with the correct api key etc.

WiiPlayer2 replied:
This only allows to have at max a single connection to any transmission server which still goes against the architecture. I suggest using the newly introduced `SimpleConnector` as base class. You can take a look at `RssConnector` or `TelegramConnector` for examples on how to use it.

itshoro committed:
Merge remote-tracking branch 'origin/dev' into transmission-feature

itshoro committed:
Added TransmissionConnector to the ConnectorRegistry

- Made TransmissionConnector public
- Removed generic type from TransmissionConnection

itshoro committed:
Change TransmissionConnector to generate a Transmission Client

- Add TransmissionConfiguration
- Generate a Transmission.API.RPC.Client based on data from TransmissionConfiguration
- Removed client parameter from the TransmissionConnector constructor, as it is no longer needed

itshoro committed:
Change TransmissonConnection to return a new connection

itshoro committed:
Merge remote-tracking branch 'origin/dev' into transmission-feature

itshoro committed:
Modified TransmissionConnector to inherit from SimpleConnector

itshoro committed:
transmission.add now adds returned data into the context

- Cleaned up TransmissionConfiguration
